nupic/research/frameworks/vernon/mixins/gradient_metrics.py

```diff
@@ -0,0 +1,232 @@
+
+
+    :param config: a dict containing the following
+
+        - gradient_metrics_args: a dict containing the following
+
+            - include_modules: (optional) a list of module types to track
+            - include_names: (optional) a list of module names to track e.g.
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+            - plot_freq: (optional) how often to create the plot, measured in training
+                         iterations; defaults to 1
+            - metrics: a list of metrics options from ["cosine", "dot",
+            "pearson"]; defaults to ["cosine",]
+            - gradient_transformation: (optional) can be one of "real", "sign",
+            "mask". "real" corresponds to the real values of the gradients,
+            "sign" corresponds to collecting the sign of the gradients, and "mask"
+            results in a binary mask corresponding to nonzero gradients; defaults to
+            "real"
+            - max_samples_to_track: (optional) how many of samples to use for plotting;
+                                   only the newest will be used; defaults to 100
+
+    Example config:
+    ```
+    config=dict(
+        gradient_metrics_args=dict(
+            include_modules=[torch.nn.Linear, KWinners],
+            plot_freq=1,
+            max_samples_to_track=150,
+            metrics=["dot", "pearson"],
+            gradient_values="mask"
+        )
+    )
+    ```
+    """
+
+    def setup_experiment(self, config):
+        super().setup_experiment(config)
+
+        # Process config args
+        gradient_metrics_args = config.get("gradient_metrics_args", {})
+        self.gradient_metrics_plot_freq, self.gradient_metrics_filter_args, \
+            self.gradient_metrics_max_samples, self.gradient_metrics, \
+            self.gradient_values = self.process_gradient_metrics_args(
+                gradient_metrics_args)
+
+        # Register hook for tracking hidden activations
+        named_modules = filter_modules(self.model, **self.gradient_metrics_filter_args)
+        hook_args = dict(max_samples_to_track=self.gradient_metrics_max_samples)
+        self.gradient_metric_hooks = ModelHookManager(
+            named_modules, TrackGradientsHook, hook_type="backward", hook_args=hook_args
+        )
+
+        # Log the names of the modules being tracked
+        tracked_names = pformat(list(named_modules.keys()))
+        self.logger.info(f"Tracking gradients for modules: {tracked_names}")
+
+        # The targets will be collected in `self.error_loss` in a 1:1 fashion
+        # to the tensors being collected by the hooks.
+        self.gradient_metric_targets = torch.tensor([]).long()
+
+    def process_gradient_metrics_args(self, gradient_metric_args):
```

Comment:
Since this method doesn't use `self`, I recommend making it a separate function in this same module.